packages/ckeditor5-engine/src/view/matcher.js

```diff
@@ -242,40 +239,156 @@ function matchName( pattern, name ) {
 // @returns {Array|null} Returns array with matched attribute names or `null` if no attributes were matched.
-function matchAttributes( patterns, element ) {
+function matchPatterns( patterns, items, getter ) {
+	const normalizedPatterns = normalizePatterns( patterns );
+	const normalizedItems = Array.from( items );
 	const match = [];
 
-	for ( const name in patterns ) {
-		const pattern = patterns[ name ];
+	normalizedPatterns.forEach( ( [ patternKey, patternValue ] ) => {
+		normalizedItems.forEach( itemKey => {
+			if (
+				isKeyMatched( patternKey, itemKey ) &&
+				isValueMatched( patternValue, itemKey, getter )
+			) {
+				match.push( itemKey );
+			}
+		} );
+	} );
 
-		if ( element.hasAttribute( name ) ) {
-			const attribute = element.getAttribute( name );
+	// Return matches only if there are at least as many of them as there are patterns.
+	// The RegExp pattern can match more than one item.
+	if ( !normalizedPatterns.length || match.length < normalizedPatterns.length ) {
+		return null;
+	}
 
-			if ( pattern === true ) {
-				match.push( name );
-			} else if ( pattern instanceof RegExp ) {
-				if ( pattern.test( attribute ) ) {
-					match.push( name );
-				} else {
-					return null;
-				}
-			} else if ( attribute === pattern ) {
-				match.push( name );
-			} else {
-				return null;
+	return match;
+}
+
+// Bring all the possible pattern forms to an array of objects with `key` and `value` property.
+// For example:
+//
+//	{
+//		src: /https:.*/
+//	}
+//
+// to:
+//
+//	{
+//		key: 'src',
+//		value: /https:.*/
+//	}
+// @param {Object|Array} patterns
+// @returns {Array|null} Returns an array of objects or null if provided patterns were not in an expected form.
+function normalizePatterns( patterns ) {
+	if ( Array.isArray( patterns ) ) {
+		return patterns.map( pattern => {
+			if ( isPlainObject( pattern ) && pattern.key && pattern.value ) {
+				return [ pattern.key, pattern.value ];
 			}
-		} else {
-			return null;
-		}
+
+			// Assume the pattern is either String or RegExp.
+			return [ pattern, true ];
+		} );
+	} else if ( isPlainObject( patterns ) ) {
+		return Object.entries( patterns );
 	}
 
-	return match;
+	// Other cases (true, string or regexp).
+	return [ [ patterns, true ] ];
+}
+
+// @param {String|RegExp} patternKey A pattern representing a key we want to match.
+// @param {String} itemKey An actual item key (e.g. `'src'`, `'background-color'`, `'ck-widget'`) we're testing against pattern.
+// @returns {Boolean}
+function isKeyMatched( patternKey, itemKey ) {
+	return patternKey === true ||
+		patternKey === itemKey ||
+		patternKey instanceof RegExp && patternKey.test( itemKey );
+}
+
+// @param {String|RegExp} patternValue A pattern representing a value we want to match.
+// TODO param
+// @param {String} itemValue An actual item value (e.g. `'http://example.com/'`, `'red'`) we're testing against pattern.
+// @returns {Boolean}
+function isValueMatched( patternValue, itemKey, getter ) {
```

Comment:
`getter` is not documented. Also, maybe it could be renamed to `attributeGetter` or some other more verbose name?